test/resources/oauth_proxy_auth.go

```diff
@@ -0,0 +1,250 @@
+	"net/http"
+	"net/http/cookiejar"
+	"net/http/httputil"
+	"net/url"
+)
+
+// Login a user through the oauth proxy
+func ProxyOAuth(host string, username string, password string) (*http.Client, error) {
+	// Create the http client with a cookie jar
+	j, err := cookiejar.New(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initalize the cookie jar: %s", err)
+	}
+
+	transport := http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+
+	client := &http.Client{
+		Jar:           j,
+		Transport:     &transport,
+		CheckRedirect: func(_ *http.Request, _ []*http.Request) error { return nil },
+	}
+
+	// Start the authentication
+	u := fmt.Sprintf("%s/oauth/start", host)
+	response, err := client.Get(u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open %s: %s", u, err)
+	}
+	if response.StatusCode != 200 {
+		return nil, errorWithResponseDump(response, fmt.Errorf("the request to %s failed with code %d", u, response.StatusCode))
+	}
+
+	// Select the testing IDP
+	document, err := parseResponse(response)
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	// find the link to the testing IDP
+	link, err := findElement(document, fmt.Sprintf("a:contains('%s')", testingIDP))
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	// get the url from the
+	href, err := getAttribute(link, "href")
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	u, err = resolveRelativeURL(response, href)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err = client.Get(u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to request %s: %s", u, err)
+	}
+	if response.StatusCode != 200 {
+		return nil, errorWithResponseDump(response, fmt.Errorf("the request to %s failed with code %d", u, response.StatusCode))
+	}
+
+	// Submit the username and password
+	document, err = parseResponse(response)
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	// find the form for the login
+	form, err := findElement(document, "#kc-form-login")
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	// retrieve the action of the form
+	action, err := getAttribute(form, "action")
+	if err != nil {
+		return nil, errorWithResponseDump(response, err)
+	}
+
+	u, err = resolveRelativeURL(response, action)
+	if err != nil {
+		return nil, err
+	}
+
+	// submit the form with the username and password
+	v := url.Values{"username": []string{username}, "password": []string{password}}
+	response, err = client.PostForm(u, v)
+	if err != nil {
+		return nil, fmt.Errorf("failed to request %s: %s", u, err)
+	}
+	if response.StatusCode != 200 {
+		return nil, errorWithResponseDump(response, fmt.Errorf("the request to %s failed with code %d", u, response.StatusCode))
+	}
+
+	document, err = parseResponse(response)
+	if err != nil {
+		return nil, err
+	}
+
+	// find the form for the approval
+	form = document.Find("form")
```

Comment:
I removed this because the URL is different when you get the approval form. But I can add that check after the approval (or if no approval is needed).